src/transformers.jl

```diff
@@ -28,20 +28,73 @@ Non-mutating version of [`transform!`](@ref).
 function transform end
 
 """
-    Transforms.apply!(data::T, ::Transform; kwargs...) -> T
+    apply(data::T, ::Transform; kwargs...) -> T
+
+Applies the [`Transform`](@ref) to the data. New transforms should usually only extend
+`_apply` which this method delegates to.
+
+Where possible, this should be extended for new data types `T`.
+"""
+function apply end
+
+"""
+    apply!(data::T, ::Transform; kwargs...) -> T
+
+Applies the [`Transform`](@ref) mutating the input `data`. New transforms should usually
+only extend `_apply!` which this method delegates to.
 
-Applies the [`Transform`](@ref) mutating the input `data`.
 Where possible, this should be extended for new data types `T`.
 """
 function apply! end
 
+
 """
-    Transforms.apply(data::T, ::Transform; kwargs...) -> T
+    apply(A::AbstractArray, ::Transform; dims=:, inds=:, kwargs...)
+
+Applies the [`Transform`](@ref) to each element of `A`.
+
+Optionally specify the `dims` to apply the [`Transform`](@ref) along certain dimensions
+and `inds` will be the indices to apply the Transform to along the `dims` specified.
 
-Non-mutating version of [`apply!`](@ref), which it delegates to by default.
-Does not need to be extended unless a mutating [`Transform`](@ref) is not possible.
+If `dims` === : (all dimensions), then `inds` will be the global indices of the array,
+instead of being relative to a certain dimension.
 """
-function apply end
+function apply(A::AbstractArray, t::Transform; dims=:, inds=:, kwargs...)
+    if dims === Colon()
+        if inds === Colon()
+            return _apply(A, t; kwargs...)
+        else
+            if A isa KeyedArray
+                # KeyedArrays don't support indexing into them via an array of indices
+                return [_apply(A[ind], t; kwargs...) for ind in inds]
+            else
+                # Apply to global indices `inds`, not `inds` relative to a certain dimension
```

Comment:
Just working this through myself. 

`inds` with `dim=:`, vectorises the elements of an array

So `apply(M, t; inds=[2,3])` yields

```julia
julia> M = [1 2 3; 4 5 6; 7 8 9];

julia> inds = [2, 3];

julia> M[inds]
2-element Array{Int64,1}:
 4
 7
```

Is this an intuitive result?
From our POV it is because we see what `dims=:` is doing. 
But will a new user have the same understanding?

Is it better to default `dims=1` instead? so that `apply(M, t; inds=[2,3])` gives
```julia
julia> M[:, inds]
3×2 Array{Int64,2}:
 2  3
 5  6
 8  9
```

I'm not suggesting this way is better, I'm fine with this way too.
 I'm just trying to imagine what a non-developer would think when trying `apply(M, t; inds=[2,3])` for the first time.

I guess the docs/examples we can show a user what to expect.

Reply:
I don't really have any pre-existing intuition around `dims` or operations on matrices so I can't really say

Reply:
Cool. Let's leave it as is.